src/Swarm/Game/State.hs

```diff
@@ -162,11 +166,34 @@ data RunStatus
     AutoPause
   deriving (Eq, Show)
 
+-- | Status of the scenario goal: whether there is one, and whether it has been
+--   displayed to the user initially.
+data GoalStatus
+  = -- | There is no goal.
+    NoGoal
+  | -- | There is a goal, and we should display it to the user initially.
+    UnreadGoal [Text]
+  | -- | There is a goal, and we have already displayed it to the user.
+    --   It can be displayed again if the user chooses.
+    ReadGoal [Text]
+  deriving (Eq, Show)
+
+-- | Do we need to display the goal initially to the user?
+goalNeedsDisplay :: GoalStatus -> Maybe [Text]
+goalNeedsDisplay (UnreadGoal g) = Just g
+goalNeedsDisplay _ = Nothing
+
+-- | Mark the goal as having been read.
+markGoalRead :: GoalStatus -> GoalStatus
+markGoalRead (UnreadGoal g) = ReadGoal g
+markGoalRead gs = gs
+
 -- | The main record holding the state for the game itself (as
 --   distinct from the UI).  See the lenses below for access to its
 --   fields.
 data GameState = GameState
   { _creativeMode :: Bool
+  , _gameGoal :: GoalStatus
```

Comment:
Out of curiosity, why do we keep this information in GameState and not the UI state?

It does not seem to interact with the game logic at all.

My guess is it's because of the YAML --> Scenario --> GameState loading, but surely that should be possible to tweak.

Reply:
Yes, that's the simplest reason why. But yes, maybe the `playScenario` function should update the entire `AppState`, not just the `GameState`.  Come to think of it that might be a better place to put the code from #445 .  I will look into that.